### PR TITLE
Delay kubeconfig loading to the filter method.

### PR DESCRIPTION
**Pull request by andriy-dmytruk**
This is required, since in some cases kubeconfig might be loaded with a client. But since client cannot be created before all filters are initialized, a circular dependency may occur.

@yawkat Do you know if there is a better way of doing this?

Also, in my case for some reason circular dependency was not detected, but instead this was printed endlessly. Is there some special filter logic that causes this?
```
Path Taken: new KubernetesHttpClientFactory(KubeConfigLoader kubeConfigLoader,KubernetesClientConfiguration kubernetesClientConfiguration,KubernetesPrivateKeyLoader kubernetesPrivateKeyLoader,ResourceResolver resourceResolver,DefaultHttpClientFilterResolver defaultHttpClientFilterResolver) --> new KubernetesHttpClientFactory(KubeConfigLoader kubeConfigLoader,KubernetesClientConfiguration kubernetesClientConfiguration,KubernetesPrivateKeyLoader kubernetesPrivateKeyLoader,ResourceResolver resourceResolver,[DefaultHttpClientFilterResolver defaultHttpClientFilterResolver]) --> ContainerEngineClientFactory.setProvider([HttpProvider provider]) --> new ManagedNettyHttpProvider([HttpClientRegistry<HttpClient T> mnHttpClientRegistry],ExecutorService ioExecutor,ObjectMapper jsonMapper,OciSerdeConfiguration ociSerdeConfiguration,OciSerializationConfiguration ociSerializationConfiguration,List<OciNettyClientFilter<Object R> E> nettyClientFilters)
Path Taken: new KubernetesHttpClientFactory(KubeConfigLoader kubeConfigLoader,KubernetesClientConfiguration kubernetesClientConfiguration,KubernetesPrivateKeyLoader kubernetesPrivateKeyLoader,ResourceResolver resourceResolver,DefaultHttpClientFilterResolver defaultHttpClientFilterResolver) --> new KubernetesHttpClientFactory(KubeConfigLoader kubeConfigLoader,KubernetesClientConfiguration kubernetesClientConfiguration,KubernetesPrivateKeyLoader kubernetesPrivateKeyLoader,ResourceResolver resourceResolver,[DefaultHttpClientFilterResolver defaultHttpClientFilterResolver]) --> ContainerEngineClientFactory.setProvider([HttpProvider provider]) --> new ManagedNettyHttpProvider([HttpClientRegistry<HttpClient T> mnHttpClientRegistry],ExecutorService ioExecutor,ObjectMapper jsonMapper,OciSerdeConfiguration ociSerdeConfiguration,OciSerializationConfiguration ociSerializationConfiguration,List<OciNettyClientFilter<Object R> E> nettyClientFilters)
Path Taken: new KubernetesHttpClientFactory(KubeConfigLoader kubeConfigLoader,KubernetesClientConfiguration kubernetesClientConfiguration,KubernetesPrivateKeyLoader kubernetesPrivateKeyLoader,ResourceResolver resourceResolver,DefaultHttpClientFilterResolver defaultHttpClientFilterResolver) --> new KubernetesHttpClientFactory(KubeConfigLoader kubeConfigLoader,KubernetesClientConfiguration kubernetesClientConfiguration,KubernetesPrivateKeyLoader kubernetesPrivateKeyLoader,ResourceResolver resourceResolver,[DefaultHttpClientFilterResolver defaultHttpClientFilterResolver]) --> ContainerEngineClientFactory.setProvider([HttpProvider provider]) --> new ManagedNettyHttpProvider([HttpClientRegistry<HttpClient T> mnHttpClientRegistry],ExecutorService ioExecutor,ObjectMapper jsonMapper,OciSerdeConfiguration ociSerdeConfiguration,OciSerializationConfiguration ociSerializationConfiguration,List<OciNettyClientFilter<Object R> E> nettyClientFilters)
Path Taken: new KubernetesHttpClientFactory(KubeConfigLoader kubeConfigLoader,KubernetesClientConfiguration kubernetesClientConfiguration,KubernetesPrivateKeyLoader kubernetesPrivateKeyLoader,ResourceResolver resourceResolver,DefaultHttpClientFilterResolver defaultHttpClientFilterResolver) --> new KubernetesHttpClientFactory(KubeConfigLoader kubeConfigLoader,KubernetesClientConfiguration kubernetesClientConfiguration,KubernetesPrivateKeyLoader kubernetesPrivateKeyLoader,ResourceResolver resourceResolver,[DefaultHttpClientFilterResolver defaultHttpClientFilterResolver]) --> ContainerEngineClientFactory.setProvider([HttpProvider provider]) --> new ManagedNettyHttpProvider([HttpClientRegistry<HttpClient T> mnHttpClientRegistry],ExecutorService ioExecutor,ObjectMapper jsonMapper,OciSerdeConfiguration ociSerdeConfiguration,OciSerializationConfiguration ociSerializationConfiguration,List<OciNettyClientFilter<Object R> E> nettyClientFilters)
...
```

### CODE REVIEW

**Review by msupic — APPROVED** [review]

File: kubernetes-client-openapi-common/src/main/java/io/micronaut/kubernetes/client/openapi/KubernetesHttpClientFilter.java [hunk 1]
```diff
@@ -40,18 +40,24 @@
 @Internal
 final class KubernetesHttpClientFilter {
 
-    private final KubeConfig kubeConfig;
+    private final KubeConfigLoader kubeConfigLoader;
+    private KubeConfig kubeConfig;
+    private boolean retrievedConfig = false;
```
Comment (yawkat) [inline kubernetes-client-openapi-common/src/main/java/io/micronaut/kubernetes/client/openapi/KubernetesHttpClientFilter.java]:
```suggestion
    private volatile boolean retrievedConfig = false;
```

**Comment by yawkat** [comment]
You could use a SupplerUtil.memoized. but I think the general idea is sound, we can't use the client to set up a client filter.

**Comment by msupic** [comment]
@andriy-dmytruk - I tried to create a test which reproduce the issue: https://github.com/micronaut-projects/micronaut-kubernetes/compare/6.2.x...client-config-loader-test but it doesn't fail. Could you please modify it to reproduce the issue and add to this PR?

**Comment by andriy-dmytruk** [comment]
> @andriy-dmytruk - I tried to create a test which reproduce the issue: [6.2.x...client-config-loader-test](https://github.com/micronaut-projects/micronaut-kubernetes/compare/6.2.x...client-config-loader-test) but it doesn't fail. Could you please modify it to reproduce the issue and add to this PR?

Finally reproduced it. You have to actually inject the client

**Comment by andriy-dmytruk** [comment]
The solution had to be even more complex than I expected. Turns out that `CredentialLoader` injects `KubeConfigLoader`, so they had to be delayed to runtime, as well.

File: kubernetes-client-openapi-common/src/main/java/io/micronaut/kubernetes/client/openapi/KubernetesHttpClientFilter.java [hunk 2]
```diff
@@ -40,18 +43,23 @@
 @Internal
 final class KubernetesHttpClientFilter {
 
-    private final KubeConfig kubeConfig;
-    private final List<KubernetesTokenLoader> kubernetesTokenLoaders;
+    private Provider<KubeConfig> kubeConfigProvider;
```
Comment (msupic) [inline kubernetes-client-openapi-common/src/main/java/io/micronaut/kubernetes/client/openapi/KubernetesHttpClientFilter.java]: Can `kubeConfigProvider` be `final`?

**Review by msupic — APPROVED** [review]

**Comment by andriy-dmytruk** [comment]
@msupic Can we release this?

**Pull request merged**